src/main/java/com/commercetools/sync/products/utils/ProductSyncUtils.java

```diff
@@ -104,7 +114,69 @@
         // lastly publish/unpublish product
         buildPublishUpdateAction(oldProduct, newProduct).ifPresent(updateActions::add);
 
-        return updateActions;
+        return prioritizeUpdateActions(updateActions,
+                oldProduct.getMasterData().getStaged().getMasterVariant().getId());
+    }
+
+    /**
+     *
+     * @param updateActions All generated update actions for all variants
+     * @param oldMasterVariantId The masterVariant of the old product fetched from Target
+     * @return An ordered list of UpdateActions as follows:
+     *              1 removeVariant actions except master
+     *              2 sameForAll Actions before executing addVariant to avoid possible errors
+     *              3 addVariant actions
+     *              4 changeMasterVariant if any
+     *              5 removeVariant for old master
+     *              6 the rest comes in
+     */
+    private static List<UpdateAction<Product>> prioritizeUpdateActions(
+            final List<UpdateAction<Product>> updateActions, final Integer oldMasterVariantId) {
+
+        final RemoveVariant removeMasterVariantUpdateAction = RemoveVariant.ofVariantId(oldMasterVariantId);
+
+        final List<UpdateAction<Product>> removeVariantUpdateActionsNoMaster =
+                getActionsByActionName(updateActions, action -> action.getAction().equals(REMOVE_VARIANT_ACTION_NAME)
+                        && ! action.equals(removeMasterVariantUpdateAction));
+
+        final List<UpdateAction<Product>> sameForAllUpdateActions =
+                getActionsByActionName(updateActions, action ->
+                        action.getAction().equals(SET_ATTRIBUTE_IN_ALL_VARIANTS_ACTION_NAME));
+
+        final List<UpdateAction<Product>> addVariantUpdateActions =
+                getActionsByActionName(updateActions, action ->
+                        action.getAction().equals(ADD_VARIANT_ACTION_NAME));
+
+        final List<UpdateAction<Product>> changeMasterUpdateActions =
+                getActionsByActionName(updateActions, action ->
+                        action.getAction().equals(CHANGE_MASTER_VARIANT_ACTION_NAME));
+
+        final List<UpdateAction<Product>> removeOldMasterVariantUpdateAction =
+                getActionsByActionName(updateActions, action ->
+                        action.getAction().equals(REMOVE_VARIANT_ACTION_NAME)
+                                && action.equals(removeMasterVariantUpdateAction));
+
+        final List<UpdateAction<Product>> updateActionList = new ArrayList<>(removeVariantUpdateActionsNoMaster);
+        updateActionList.addAll(sameForAllUpdateActions);
+        updateActionList.addAll(addVariantUpdateActions);
+        updateActionList.addAll(changeMasterUpdateActions);
+        updateActionList.addAll(removeOldMasterVariantUpdateAction);
+        updateActionList.addAll(updateActions);
```

Comment:
I wonder why we are applying other actions after removeMaster, isn't it better after SameForAllUpdateActions?

Reply:
In order to remove the old master variant you have to add the new variant first and change it to master then the CTP will allow the removal of the old master variant as it's not a master anymore...the rest of update actions comes latest after the removal of the old master variant just to avoid applying any other updates to a variant that is meant to be removed later.
Hope that this answers your question.
Thanks